File: app/javascript/channels/community_feed_channel.js
import consumer from "channels/consumer"

// Subscribe to CommunityFeedChannel when the page loads
// This ensures presence tracking is activated
let subscription = null
let heartbeatInterval = null

document.addEventListener("turbo:load", () => {
  // Only subscribe if we're on the community page
  if (document.querySelector('[data-controller*="community-composer-modal"]')) {
    if (!subscription) {
      subscription = consumer.subscriptions.create("CommunityFeedChannel", {
        connected() {
          console.log("Connected to CommunityFeedChannel - presence tracking active")
          
          // Send heartbeat every 30 seconds to keep presence alive
          // (cache TTL is 45 seconds)
          heartbeatInterval = setInterval(() => {
            if (this.consumer.connection.isActive()) {
              this.perform("heartbeat")
            }
          }, 30000)
        },

        disconnected() {
          console.log("Disconnected from CommunityFeedChannel")
          
          if (heartbeatInterval) {
            clearInterval(heartbeatInterval)
            heartbeatInterval = null
          }
        },

        received(data) {
          // Handle any custom messages from the channel if needed
          console.log("Received:", data)
        }
      })
    }
  }
})

// Clean up subscription when leaving the page
document.addEventListener("turbo:before-cache", () => {
  if (subscription) {
    subscription.unsubscribe()
    subscription = null
  }
  
  if (heartbeatInterval) {
    clearInterval(heartbeatInterval)
    heartbeatInterval = null
  }
})

File: app/channels/community_feed_channel.rb
class CommunityFeedChannel < ApplicationCable::Channel
  def subscribed
    stream_from "community_feed"
    track_presence!
  end

  def unsubscribed
    Community::PresenceTracker.remove(visitor_uid)
    Community::PresenceTracker.broadcast!
  end

  def heartbeat
    # Keep presence alive by updating last_seen_at timestamp
    track_presence!
  end

  private

  def track_presence!
    Community::PresenceTracker.touch(
      visitor_uid,
      name: visitor_name
    )
    Community::PresenceTracker.broadcast!
  end
end

